Account/FxADeviceRegistrator.swift

```diff
@@ -0,0 +1,116 @@
+        if let _ = account.fxaDeviceId where account.deviceRegistrationVersion == DeviceRegistrationVersion {
+            return deferMaybe(FxADeviceRegistrationResult.AlreadyRegistered)
+        }
+
+        let client = FxAClient10(endpoint: account.configuration.authEndpointURL)
+        let sessionToken = state.sessionToken
+        let name = DeviceInfo.defaultClientName()
+        let device: FxADevice
+        let registrationResult: FxADeviceRegistrationResult
+        if let deviceId = account.fxaDeviceId {
+            device = FxADevice.forUpdate(name, id: deviceId)
+            registrationResult = FxADeviceRegistrationResult.Updated
+        } else {
+            device = FxADevice.forRegister(name, type: "mobile")
+            registrationResult = FxADeviceRegistrationResult.Registered
+        }
+
+        let registeredDevice = client.registerOrUpdateDevice(sessionToken, device: device)
+        let deviceId: Deferred<Maybe<String>> = registeredDevice.bind { result in
+            if let device = result.successValue {
+                return deferMaybe(device.id!)
+            }
+
+            // Recover from the error -- if we can.
+            if let error = result.failureValue as? FxAClientError,
+               case .Remote(let remoteError) = error {
+                switch (remoteError.code) {
+                case FxAccountRemoteError.DeviceSessionConflict:
+                    return recoverFromDeviceSessionConflict(account, client: client, sessionToken: sessionToken)
+                case FxAccountRemoteError.InvalidAuthenticationToken:
+                    return recoverFromTokenError(account, client: client)
+                default: break
+                }
+            }
+
+            // Not an error we can recover from. Rethrow it and fall back to the failure handler.
+            return deferMaybe(result.failureValue!)
+        }
+
+        // Post-recovery. We either got the device ID or we didn't, but update the account either way.
+        return deviceId.bind { result in
+            switch result {
+            case .Success(let deviceId):
+                account.fxaDeviceId = deviceId.value
+                account.deviceRegistrationVersion = DeviceRegistrationVersion
+                return deferMaybe(registrationResult)
+            case .Failure(let error):
+                log.error("Device registration failed: \(error.description)")
+                account.fxaDeviceId = nil
+                account.deviceRegistrationVersion = 0
+                return deferMaybe(error)
+            }
+        }
+    }
+
+    private static func recoverFromDeviceSessionConflict(account: FirefoxAccount, client: FxAClient10, sessionToken: NSData) -> Deferred<Maybe<String>> {
+        log.warning("Device session conflict. Attempting to find the current device ID…")
+        return client.devices(sessionToken) >>== { response in
+            guard let currentDevice = response.devices.find({ $0.isCurrentDevice! }) else {
```

Comment:
Force unwrap could be dangerous but as mentioned above I think we can make this non-optional and just default to false.